pkg/util/timeout.go

```diff
@@ -0,0 +1,37 @@
+package util
+
+import (
+	"fmt"
+	"time"
+)
+
+// TimeoutError is error returned after timeout occured.
+type TimeoutError struct {
+	after time.Duration
+}
+
+// Error implements the Go error interface.
+func (t *TimeoutError) Error() string {
+	return fmt.Sprintf("calling the function timeout after %v", t.after)
+}
+
+// TimeoutAfter executes the provide function and return the TimeoutError in
+// case when the execution time of the provided function is bigger than provided
+// time duration.
+func TimeoutAfter(t time.Duration, fn func() error) error {
+	c := make(chan error, 1)
+	defer close(c)
```

Comment:
The buffer in the channel is the right pattern to avoid leaking a goroutine, well done :)

There's still a bug here. When we time out, the goroutine started in the next line will panic sending on a closed channel.

It should be:

```go
c := make(chan error, 1)
go func() {
  defer close(c)
  c <- fn()
}()
select {
```

That ensures the channel is only closed after we send.